plugins/regex_revalidate/regex_revalidate.cc

```diff
@@ -0,0 +1,729 @@
+  TSHttpTxnReenable(txn, TS_EVENT_HTTP_CONTINUE);
+  return 0;
+}
+
+} // namespace
+
+namespace
+{
+void
+setup_memory_allocation()
+{
+  if (&ts_malloc != pcre_malloc) {
+    pcre_malloc = &ts_malloc;
+    pcre_free   = &ts_free;
+  }
+}
+
+void
+setup_config_cont(PluginState *const pstate)
+{
+  if (nullptr == config_cont) {
+    DEBUG_LOG("creating config continuation");
+
+    // create reusable configuration file regex
+    constexpr char const *const expr = "^([^#].+?)\\s+(\\d+)\\s*$";
+    bool const cstat                 = config_re.compile(expr);
+
+    if (!cstat) {
+      ERROR_LOG("setup_config_cont: Unable to compile config_re");
+      return;
+    }
+
+    // set up the config continuation
+    config_cont = TSContCreate(config_handler, TSMutexCreate());
+    TSContDataSet(config_cont, static_cast<void *>(pstate));
+
+    // constantly run this for rule grooming
+    TSContScheduleOnPool(config_cont, CONFIG_TMOUT, TS_THREAD_POOL_TASK);
+
+    // also register this continuation with any config reloads
+    TSMgmtUpdateRegister(config_cont, PLUGIN_NAME);
+
+    // and also register as a lifecycle hook
+    TSLifecycleHookAdd(TS_LIFECYCLE_MSG_HOOK, config_cont);
+  }
+}
+
+} // namespace
+
+void
+TSPluginInit(int argc, char const *argv[])
+{
+  DEBUG_LOG("Starting plugin init");
+
+  TSPluginRegistrationInfo info;
+  info.plugin_name   = PLUGIN_NAME;
+  info.vendor_name   = "Apache Software Foundation";
+  info.support_email = "dev@trafficserver.apache.org";
+
+  if (TSPluginRegister(&info) != TS_SUCCESS) {
+    ERROR_LOG("Global plugin registration failed");
+    return;
+  } else {
+    DEBUG_LOG("Global plugin registration succeeded");
+  }
+
+  if (TS_VERSION_MAJOR != TSTrafficServerVersionGetMajor()) {
+    ERROR_LOG("Plugin requires Traffic Server %d", TS_VERSION_MAJOR);
+    return;
+  }
+
+  setup_memory_allocation();
+
+  PluginState *const pstate = new PluginState;
+  if (!pstate->fromArgs(argc, argv)) {
+    ERROR_LOG("Remap plugin registration failed");
+    delete pstate;
+    return;
+  }
+
+  setup_config_cont(pstate);
+
+  TSCont const main_cont = TSContCreate(main_handler, nullptr);
+  TSContDataSet(main_cont, static_cast<void *>(pstate));
+  TSHttpHookAdd(TS_HTTP_CACHE_LOOKUP_COMPLETE_HOOK, main_cont);
+
+  TSAssert(nullptr != config_cont);
+
+  TSMutex const config_mutex = TSContMutexGet(config_cont);
+  TSMutexLock(config_mutex);
+
+  time_t const timenow = time(nullptr);
+
+  std::vector<Invalidate> newrules;
+  pstate->loadConfig(timenow, &newrules);
+  pstate->config_file_mtime = timeForFile(pstate->config_file);
+
+  time_t new_expiry = 0;
+  pruneConfig(timenow, &newrules, &new_expiry);
+  pstate->min_expiry = new_expiry;
+
+  auto vecnew = std::make_shared<std::vector<Invalidate>>(std::move(newrules));
+
+  std::atomic_store(&(pstate->invalidate_vec), vecnew);
+  DEBUG_LOG("Configuation installed");
+
+  pstate->logConfig();
+
+  TSMutexUnlock(config_mutex);
+
+  DEBUG_LOG("Global Plugin Init Complete");
+}
```

Comment:
I notice that the keyword 'private' is not used in this file.  Is there really no value in making any of the members of any of the classes private?

Reply:
It's a struct, not a class.